### DISCUSSION TITLE
IC-R8600

### DISCUSSION THREAD

**Discussion by scancapecod** [AlexandreRouma/SDRPlusPlus#268] [Q&A]
Hello,

Just stumbled across this software today and I must say I'm impressed out of the gate. I'm currently using it with an Airspy R2 and an Airspy Mini. I also own an Icom IC-R8600. Is there a way to make it work with SDRpp? It seems there is some soapy support for the 8600 and I believe there is soapy support in SDRpp, but I just don't know how to tie the two together or if it is even possible. I'm using Windows in all cases. 

Thanks in advance, and thanks for this awesome software.

Scott/KC1UA

**Comment by AlexandreRouma** [comment]
To use SoapySDR with SDR++, you need to install PothosSDR (make sure to enable the "add to path" option in the installer)
You'll then see a "SoapySDR" option in the source list

**Comment by scancapecod** [comment]
Thanks for the response. I'll whittle away at this when I get the chance. I'm looking forward to seeing how this software progresses. Thanks for your efforts!

**Comment by scancapecod** [comment]
Just a followup....I have managed to make SDR++ Windows X64 version work with the IC-R8600. To do so I originally got the 8600 working with another SDR program, SDRGlut. Then, I copied the following files from SDRGlut over to SDR++ 

SoapySDR.dll
SoapySDRUtil.exe
ExtIO_ICR8600.dll
icr8600Support.dll was copied into the "Modules" folder

On startup of SDR++ (previously not run on the PC I'm on) it immediately identified SoapySDR as the source, and underneath was a drop down window with IC-R8600 in it as the only choice. I had to make several tweaks to the attenuator gain, RFGain, bandwidth, and adjust the max and min display settings. Clicked the start button and lo and behold it works. 

There are a few quirks but it is a great start. I'm noticing that at a bandwidth of 5.12 MHz (the maximum for the 8600) I am truly seeing about 4.5 MHz. Also, strong signals appear to make the spectrum "jump" for lack of a better term. I've tried attenuation and reducing RF gain and it minimizes, but does not eliminate, this issue. Also. When I stop the spectrum display SDR++ shuts down and thus requires a restart to continue. On the spectrum display the bandwidth does not fill the entire screen regardless of what it is set at, 5.12, 3.84, 1.92. I did not go lower during my initial test runs.

All of that said, it does work! More to follow as I try it further. Thanks again for this software!

Scott/KC1UA

**Reply by AlexandreRouma** [reply]
Hello,

This is not the recommended way of getting SoapySDR running with SDR++ and will cause issues for most users (in some cases completely breaking SDR++ and/or SoapySDR). The soapysdr module for the IC8600 doesn't seem to have been compiled for the recent versions of SoapySDR so granted you can't do much better at the moment. But just keep this in mind

I'll see if I can compile that driver for you so you can install it properly

**Comment by scancapecod** [comment]
Thanks Alexandre. I'm sure it's not the right way to do it! Thanks for the response. If you can compile a driver that would be fantastic and greatly appreciated.

**Reply by AlexandreRouma** [reply]
Alright, I've successfully compiled it.

So, first, I'd suggest backing up your current SDR++ and archiving it in case this doesn't work for you:

1) Install SDR++ in a clean location
2) Download and install [PothosSDR](https://downloads.myriadrf.org/builds/PothosSDR/), making sure to check the "Add to path" option in the installer.
3) Download and extract [icr8600Support.dll](https://github.com/AlexandreRouma/SDRPlusPlus/files/7107712/icr8600Support.zip) (make sure it's that one, the one you used is meant for older SoapySDR versions) to the following directory: `C:\Program Files\PothosSDR\lib\SoapySDR\modules0.8`
4) Start SDR++ and select "SoapySDR" as a source, you should see your radio and be able to use it like usual

Regards,
Alex

**Comment by scancapecod** [comment]
Wow I didn't expect it that fast! Following your instructions and it works just fine. The only thing I'm noticing thus far is that when I stop sampling the program exits. It does not do that with my Airspy setup. I stop it to change bandwidths as that can't be done while it is sampling. I'll continue to poke around in there and see how it goes, but the method above is an absolute solution. Thanks!

**Reply by AlexandreRouma** [reply]
Did you have that same issue with the way you did it earlier?

**Comment by scancapecod** [comment]
Yes, I was seeing that behavior with my original setup as well.

**Reply by AlexandreRouma** [reply]
Alright, I'd suggest trying it with some other software supporting SoapySDR. If it also happens with another software then I think opening up an issue [on the page of the developer of the driver](https://github.com/dmitryelj/SoapyICR8600/issues) would be a good idea.

If the issue however is in SDR++ I'm afraid I won't be able to debug it myself, so someone with the skills to debug SDR++ and who owns an IC-R8600 will be needed

**Reply by righthalfplane** [reply]
@AlexandreRouma @scancapecod 

I tracked down the problem. It was in the soapy_source routine "stop"
        ...
        _this->running = false;
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
        _this->dev->deactivateStream(_this->devStream);
      ...
You need to sleep a short time giving the other thread a chance to actually finish buffering the data and stop, before continuing with the shutdown.

The HackRF One also required the pause to keep from crashing on the stop.
 As a side note - the HackRF One runs using soapysdr on the Mac and Window 10, but  if you directly select the HackRF one, sdrpp crashes immediatly trying to start it up.

**Reply by AlexandreRouma** [reply]
> You need to sleep a short time giving the other thread a chance to actually finish buffering the data and stop, before continuing with the shutdown.

No, this is a hacky way of getting around bugs in the driver. The `deactivateStream` MUST terminate the read operation which will genereally just return either no or less samples than expected to the worker. Not doing so and instead crashing is just the soapy driver not following the soapy spec.

> The HackRF One also required the pause to keep from crashing on the stop.
As a side note - the HackRF One runs using soapysdr on the Mac and Window 10, but if you directly select the HackRF one, sdrpp crashes immediatly trying to start it up.

This is unrelated. As mentioned in the troubleshooting section of the readme, a bug in libhackrf makes it impossible to have both the soapysdr_source and hackrf_source modules. You need to completely disable the soapy_source module by moving it out of the module folder (or renaming it to have a different extension than .dll or .so).

There is also another bug with libhackrf that occurs on the latest libusb version, not sure if they fixed that one yet.

**Reply by righthalfplane** [reply]
The drivers of full of bugs - If you want to use them you need to work around the bugs or fix the bugs. The standard soapy driver for  IC-R8600 pays no attention the the amount of data requested. It returns three different amounts of data - no matter what is requested - often times exceeding the amount requested - which is nice for over running buffers - that is why I have my own patched version. The standard soapy driver for the NetSDR, also, always returned the same amount - over running buffers - until I patched it. The  standard soapy driver for the RedPitaya has several small bugs that cause problems for programs that are expecting things like a "Label" or "MTU" information. When I get the time, I will forward some more fixes to Josh Blum.

**Comment by scancapecod** [comment]
Thanks. I'll see what I can come up with. Many thanks for getting it to this point though; the shutdown is a minor issue (for me anyway) and it seems to be working well other than that. Audio recovery could be a little cleaner but I believe it was posted elsewhere that was being worked on. All in all it is a great program with a lot of nice features, and is appreciated. Thanks for your time and efforts!

Scott

**Comment by scancapecod** [comment]
On another note entirely regarding the 8600...is there any possibility of eventual control via its LAN port, as opposed to using USB?

Thanks,
Scott

**Reply by AlexandreRouma** [reply]
I'm not familiar with it or its soapysdr driver but it's not currently possible to enter device arguments manually. This means even if the driver supports it, if you need to enter the IP, it's not possible yet.

**Reply by scancapecod** [reply]
Understood and thanks.
